pennylane/_grad.py

```diff
@@ -180,6 +226,73 @@ def new_func(*args, **kwargs):
     return new_func
 
 
+def _validate_cotangents(cotangents, out_avals):
+    from jax._src.api import _dtype  # pylint: disable=import-outside-toplevel
+
+    def get_shape(x):
+        return x.shape if hasattr(x, "shape") else jax.numpy.shape(x)
+
+    if len(cotangents) != len(out_avals):
+        raise ValueError(
+            "The length of cotangents must match the number of"
+            " outputs of the function with qml.vjp."
+        )
+    for p, t in zip(cotangents, out_avals):
+        if _dtype(p) != _dtype(t):
+            raise TypeError(
+                "function output params and cotangents arguments to qml.vjp do not match; "
+                "dtypes must be equal. "
+                f"Got function output params dtype {_dtype(p)} and so expected cotangent dtype "
+                f"{_dtype(p)}, but got cotangent dtype {_dtype(t)} instead."
+            )
+
+        if get_shape(p) != get_shape(t):
+            raise ValueError(
+                "qml.vjp called with different function output params and cotangent "
+                f"shapes; got function output params shape {get_shape(p)} and cotangent shape "
+                f"{get_shape(t)}"
+            )
+
+
+def _capture_vjp(func, params, cotangents, *, argnums=None, method=None, h=None):
+    from jax.tree_util import tree_leaves, tree_unflatten  # pylint: disable=import-outside-toplevel
+
+    if argnums is None:
+        argnums = [0]
+    elif isinstance(argnums, int):
+        argnums = [argnums]
```

Comment:
For some reason, jacobian has squeezes the output if the argnums is an int, but vjp doesn't. But just sticking with the same catalyst behaviour for now.